cmd/machine-config-controller/start.go

```diff
@@ -90,16 +94,16 @@ func runStartCmd(cmd *cobra.Command, args []string) {
 		go draincontroller.Run(5, ctrlctx.Stop)
 
 		// wait here in this function until the context gets cancelled (which tells us whe were being shut down)
-		<-ctx.Done()
-		glog.Infof("Context has returned done, waiting 10 seconds to let any in-progress work finish")
-		// TODO(jkyros): someday it would be nice to have this be event-driven so the controller could track and acknowledge that it had completed
-		// any in-progress work (e.g. it is in the middle of a sync), but for now just give it a little bit of time.
-		<-time.After(10 * time.Second)
+		<-runContext.Done()
+		glog.Infof("Context has returned done, waiting 5 seconds to let any in-progress work finish")
+		// TODO(jkyros): someday it might be nice to have the controller wait for any in-progress work (e.g. it is in the middle of a sync at this very moment),
+		// but for now just give it a little bit of time.
+		<-time.After(5 * time.Second)
```

Comment:
Yeah,  this was my concern too. I talked about it with Jerry a little last week. 

 My thought here was: 
- if there are no more events coming in, and no workers working (because the context cancelled them all),  `syncMachineConfigPool` can't run again because there's no event to trigger it
- so the only thing we *should* have to worry about is the one `syncMachineConfigPool` that is already in progress

but yes it would make me feel safer if we had something to stop us if some race-contion-y thing happened and something snuck through the queue while everything was reacting to the context cancellation.

Reply:
Jerry and I talked some more about this. As for that variable, checking the context should be just as good, (if cancelled, quietly no-op), but say I add a context check at the beginning of `syncMachineConfigPool`:
-  that doesn't save us from a "half-done" sync (like if we were already in the middle of the function body)
-  if we wait for that "half-done" sync, our syncs _could_ be even slower than they are today?

Reply:
Some other options we discussed:
- have the next controller, after acquiring the lease, add a small wait to ensure the previous pod is (most likely) terminated
- somehow track all the subcontrollers and routines and terminate them synchronously
- check what other controllers are doing for this

Reply:
I agree we should check what other operators are doing here.

That said, to avoid races I think the general pattern is:

- Use a "done" channel we can poll in each controller sync loop
- [Use a sync.WaitGroup](https://pkg.go.dev/sync#WaitGroup) to wait for the completion of the polling goroutines

Reply:
I did read through the [CVO PR](https://github.com/openshift/cluster-version-operator/pull/424) you sent me, the CVO is really meticulous about it (see: https://github.com/openshift/cluster-version-operator/blob/master/pkg/start/start.go), taking an approach where they: 

- explicitly try to handle crashes in and reap all their goroutines and 
- have a graceful shutdown period (2 minutes) where they allow some time for that to happen 

I don't know how much "return on investment" we get from doing that if we don't really want to wait for things to shutdown, but if we think that's the approach we want to take, I can set ours up that way also?

Reply:
Hmm, so in theory, the longest "sync" we have takes 10 minutes, which is the waiting for required pools loop. Now I think about it, if we do wait for "graceful shutdown" during upgrades, it will never happen for the first MCC pod since the required pools will never be ready in that 10 minute blocking poll.

Put another way, maybe a better way to approach this is to rework that polling loop to not be blocking, such that we could gracefully shutdown

Reply:
For correctness, everything that's making a blocking call should take a context and be cancelled when that happens.  So given your example of `syncRequiredMachineConfigPools()` for example...we should be passing a `context` to that, then it gets passed down into the API calls like 
```
diff --git a/pkg/operator/sync.go b/pkg/operator/sync.go
index 69e4aa742..dd131f751 100644
--- a/pkg/operator/sync.go
+++ b/pkg/operator/sync.go
@@ -684,7 +684,7 @@ func (optr *Operator) syncRequiredMachineConfigPools(_ *renderConfig) error {
 				return false, nil
 			}
 			optr.setOperatorStatusExtension(&co.Status, lastErr)
-			_, err = optr.configClient.ConfigV1().ClusterOperators().UpdateStatus(context.TODO(), co, metav1.UpdateOptions{})
+			_, err = optr.configClient.ConfigV1().ClusterOperators().UpdateStatus(ctx), co, metav1.UpdateOptions{})
 			if err != nil {
 				lastErr = errors.Wrapf(lastErr, "failed to update clusteroperator: %v", err)
 				return false, nil
```

Reply:
And there is also [PollWithContext](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollWithContext) instead of plain `wait.Poll` that we should use.

Reply:
How about this as a short term plan:

As soon we lose the lease, call `os.Exit(0)` - do not pass Go¹, do not collect $200.

This means we could interrupt some of the controllers mid change, but they already need to be robust to that in theory.



¹ Get it?  We're not "passing" any more Go code, the process is exiting...

Reply:
I can't say for sure myself...possibly/probably? Immediately calling `os.Exit(0)` would reinforce that that's what we expect, and be robust to someone coming along later and adding something after the leader election loss (as unlikely as that is probably).

(I don't have a really strong opinion on this bit)

But I do think we should not call `time.Sleep()` - we should exit as fast as possible.